controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -35,21 +35,24 @@ type etcdClientGenerator struct {
 	tlsConfig  *tls.Config
 }
 
-func (c *etcdClientGenerator) forNode(ctx context.Context, name string) (*etcd.Client, error) {
-	// This does not support external etcd.
+func (c *etcdClientGenerator) forNodes(ctx context.Context, nodes []corev1.Node) (*etcd.Client, error) {
+	endpoints := make([]string, len(nodes))
+	for i, node := range nodes {
+		endpoints[i] = staticPodName("etcd", node.Name)
+	}
+
 	p := proxy.Proxy{
-		Kind:         "pods",
-		Namespace:    metav1.NamespaceSystem, // TODO, can etcd ever run in a different namespace?
-		ResourceName: staticPodName("etcd", name),
-		KubeConfig:   c.restConfig,
-		TLSConfig:    c.tlsConfig,
-		Port:         2379, // TODO: the pod doesn't expose a port. Is this a problem?
+		Kind:       "pods",
+		Namespace:  metav1.NamespaceSystem,
+		KubeConfig: c.restConfig,
+		TLSConfig:  c.tlsConfig,
+		Port:       2379,
 	}
 	dialer, err := proxy.NewDialer(p)
 	if err != nil {
 		return nil, err
 	}
-	etcdclient, err := etcd.NewEtcdClient("127.0.0.1", dialer.DialContextWithAddr, c.tlsConfig)
+	etcdclient, err := etcd.NewEtcdClient(endpoints, dialer.DialContextWithAddr, c.tlsConfig)
```

Comment:
In Kubeadm, after connecting to etcd, we call client.Sync to get the client to adapt to the current list of member the etcd cluster recognizes, so we can get rid of discrepancies between the list of etcd endpoints (pods in this case) and the list of etcd members actually running.
should we do the same here as well?

Reply:
Probably not because the real IPs aren't accessible by the management cluster. We always have to use API server to get the pods.